File: src/main/java/mirai/chitung/plugin/core/harbor/Threshold.java
package mirai.chitung.plugin.core.harbor;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

class Threshold {

    final Map<Long, Integer> data = new ConcurrentHashMap<>();
    final int limit;

    Threshold(int limit) {
        this.limit = limit;
    }

    public void clear() {
        data.clear();
    }

    public void count(long id) {
        data.put(id, data.getOrDefault(id, 0) + 1);
    }

    public int get(long id) {
        return data.getOrDefault(id, 0);
    }

    public boolean reachLimit(long id) {
        return data.getOrDefault(id, 0) >= limit;
    }
}

File: src/main/kotlin/mirai/chitung/plugin/core/harbor/Threshold.kt
package mirai.chitung.plugin.core.harbor

import java.util.concurrent.ConcurrentHashMap

internal class Threshold(private val limit: Int) {
    val data: MutableMap<Long, Int> = ConcurrentHashMap()

    fun clear() {
        data.clear()
    }

    fun count(id: Long) {
        data[id] = data.getOrDefault(id, 0) + 1
    }

    operator fun get(id: Long): Int {
        return data.getOrDefault(id, 0)
    }

    fun reachLimit(id: Long): Boolean {
        return data.getOrDefault(id, 0) >= limit
    }
}
